ledokku-bootstrap.sh

```diff
@@ -1,39 +1,162 @@
+  RED="$(tput setaf 1)" # ${RED}
+  GREEN="$(tput setaf 2)" # ${GREEN}
+  YELLOW="$(tput setaf 3)" # ${YELLOW}
+  BLUE="$(tput setaf 123)" # ${BLUE}
+  END="$(tput setaf 7)" # ${END
+}
+
+function system-info() {
+  
+  # Finding Information about your device
+  ## Basic VPS info
+  ## Should exist in every script
+  IP="$(ifconfig | grep broadcast | awk '{print $2}')"
+  OS=$( $(compgen -G "/etc/*release" > /dev/null) && cat /etc/*release | grep ^NAME | tr -d 'NAME="' || echo "${OSTYPE//[0-9.]/}")
+}
+
+function check-whiptail() {
+  
+  # Checking if whiptail is available or not
+  if which whiptail >/dev/null; then
+      echo "${GREEN}whiptail exists${END}"
+      # Continue the script
+  else
+      echo "${RED}whiptail does not exist${END}"
+      echo "Install whiptail and re-run the script, your OS is ${OS}"
       exit
+      # As I already know the OS, I can also automate this process, but will save it for later.
+      # This is just a matter of finiding all the possible OS people might use and writing a if statement.
+  fi
+}
+
+function check-root() {
+  
+  #Check root and if not root take permissions (Some providers doesnot support passwordless sudo)
+  ## It is always better to do this.
+  ## We will not face any further issues, during any sort of compulsory sudo commands; like the case for installing plugins in Dokku or Giving permissions to our scripts
+  if [ "$(whoami)" == "root" ] ; then
+      echo "${YELLOW}Nice you are running the script as root!${END}"
+  else
+      echo "${RED}Please! Run the script with root access${END}, without root access I cannot create plugins in Dokku"
+      exit
+  fi
+}
+
+function system-update {
+
+  # Updating and Upgrading system
+  ## If dokku was installed, it will be automatically updated to the latest version
+  ## Staying up to date is always good
+  if (whiptail --title "Update and Upgrade System " --yes-button "Yes" --no-button "Skip"  --yesno "Do you wish to Update packges and Upgrade your system?" 10 60) then
+      echo "You chose to update your system"
+      # Update and skip to next step
+      echo "${YELLOW}Updating System${END}"
+      sudo dpkg --configure -a
+      sudo apt -y --purge autoremove
+      sudo apt install -f
+      sudo apt -y update
+      wait
+      echo "${YELLOW}Upgrading System${END}"
+      sudo apt -y upgrade
+      sudo apt -y autoclean
+      sudo apt -y --purge autoremove &
+      process_id=$!
+      wait $process_id
+      echo "Exit status: $?"
+  else
+      echo "${YELLOW}You chose to skip.${END}"
+      # Should skip to next step
+  fi
+}
+
+function ensure-dokku() {
+  
+  # Confirming the existance of Dokku
+  ## If exists, promt for update
+  ## If not exits Let him download.
+  if which dokku >/dev/null; then
+      echo "${GREEN}Dokku Exists${END}"
+      # Promt for update
+      if (whiptail --title "Updating Dokku" --yes-button "Update" --no-button "Skip"  --yesno "Would you like to update your Dokku?" 10 60) then
```

Comment:
I am not sure about prompting users to update dokku here. How to upgrade dokku is changing from one version to another so we do not cover all the versions with this approach. It's better that the user takes care of this manually and reads the dokku Changelog notes to see what might break during the upgrade.

Reply:
Yes, we can do that, but according to the present script, it will update to the latest version.

Reply:
While using ledokku, if we can force users to use the latest version, then we will not have any conflicts. 

> Keeping that in mind, I'm actually promoting such update requests.

Reply:
### What should we do about this? 
- Keep it to the latest and stable version for now and we added support for individual versions in the later stage.
 ** or**
- Uncomment those and make it manual(for the user to decide)

---
### My Opinion
If we can force all the ledokku users to one version, it would be easy for us to maintain, rather than leaving it to the discretion of the users.

Reply:
> If we can force all the ledokku users to one version, it would be easy for us to maintain, rather than leaving it to the discretion of the users.

Agree with this, just to make it more safe I would suggest the following changes:
- only print the upgrade message if user is not already using the latest version
- add a message saying something like this:
```
You are currently using Dokku vXXX, would you like to update Dokku to vXXX?
(in red) Please make sure you checked the dokku CHANGELOG before doing the upgrade https://github.com/dokku/dokku/releases
```

Reply:
So, you would like me to add a warning message (about the outdated version) and should no prompt the option to update within the script.

Am I catching up with you?

Reply:
Warning message so user understand that there is a risk by updating, and then ask him if he wants to update so keep the prompt

Reply:
So now, this is what I'm going to add now.

check dokku => if installed, check the version => if matched with the version we desire => we will pass it to the next command. 

If the installed versions are not matching, I will display the version used by the user and the latest dokku version with a warning message. And also provide the user with the ability to update to the latest dokku version and to also to skip this step.

Reply:
SOunds perfect 👌

Reply:
Check it and let me know if we need any changes.